### ISSUE TITLE
Callback function doesn't give the same value as the one being displayed by SciPy SLSQP

### ISSUE THREAD

**Issue by Mocca25** [scipy/scipy#11484]
Hi,

I'm working with the SLSQP method from scipy.optimize.minimize to solve constrained problems. I'm really interested in getting the objective function value and the x value (x being the vector of parameters) at each iteration. I'm using the "disp=True" and "iprint=2" options in order to see the results during the optimization, but I couldn't find how to get all these values into arrays automatically (to plot them, for example), so I'm also using a callback function which adds the x value and the objective value to 2 lists at each iteration.

I would expect this callback function to save the same values as those being displayed during the optimization, but it's not the case. I've tried printing the objective value at each function evaluation and I've seen the callback function saves the last evaluation. However, I don't understand what value is being displayed. I first thought it was the smallest evaluation, but sometimes it's not.
The difference between the callback value and the displayed value is often very little at last iteration but I've come across a case where the last callback x is different from the scipy.optimize.minimize result. In this case, I have seen another problem : the cost function is being evaluated after the last iteration (so no callback on these evaluations, but the result depends on these evaluations).

I tried to reproduce it on a smaller code :

```
import numpy as np
import matplotlib.pyplot as plt
import scipy.optimize as spo


def f(x):
    m = np.array([0.2,0.15,0.3])
    xx = (x-m)
    return ((1-xx[0])**2+100*(xx[1]-xx[0]**2)**2 + xx[2])**2 #**1 for second issue, if ftol=1e-6

### init
x0 = np.zeros(3)
Obj0 = f(x0)

### Objective function
def Obj(x):
    y = f(x)/Obj0  #  (Rosenbrock function + x2)**2
    print('Obj : ',y)
    return y


### Inequality Constraint
def Cons1(x): # srqt(x1**2+x2**2-x1*x2+3*x3**2) < C --> 1-sqrt(...)/C >0
    C = 4.3
    return 1-np.sqrt(x[0]**2+x[1]**2+3*x[2]**2-x[0]*x[1])/C

cons = {'type':'ineq',
        'fun':lambda x: Cons1(x)}

### Callback function
XX = [x0.copy()]
FF = [Obj(x0)]
def CB(x):
    global XX, FF
    XX += [x.copy()]
    FF += [Obj(x)]
    print('CB :  ',FF[-1])

    
### Bounds
boundmin = -5*np.ones_like(x0)
boundmax = 5*np.ones_like(x0)
bounds = spo.Bounds(boundmin, boundmax)

#%% OPTIMIZATION

res = spo.minimize(Obj, x0, callback=CB, method='SLSQP',constraints=[cons], options={'ftol': 1e-6, 'disp': True,'maxiter': 100, 'iprint': 2,'eps':1.e-6},bounds=bounds)


#%% Plotting the results
plt.figure()
plt.plot(FF)
```

The two first iterations give an objective value of 5e7 and 1e3 with the callback function, whereas 1 and 5.7e-2 are displayed. The constraints are verified at each evaluation (except for the one at 5e7). At iterations 3 and 4 (and maybe others), the displayed value is bigger than the callback value (so it's not the smallest value that is being displayed).
If I set the power to 1 in the f function (it changes the objective function), there is an evaluation after the last iteration.

I have the following questions : what exactly is displayed by scipy.optimize.minimize SLSQP ? Since callback values and displayed values are different, how to save the **displayed** values into arrays (or lists) ?
Why are some evaluations made after the last iteration ?

Thanks a lot for your help and answers, and for reading my long question.

P.S. : if it helps, I'm working with Scipy 1.1.0 version.

**Comment by andyfaff** [comment]
This will be a long answer. I modified your example a bit:
```
import numpy as np
import matplotlib.pyplot as plt
import scipy.optimize as spo


def f(x):
    m = np.array([0.2,0.15,0.3])
    xx = (x-m)
    return ((1-xx[0])**2+100*(xx[1]-xx[0]**2)**2 + xx[2])**2 #**1 for second issue, if ftol=1e-6

### init
x0 = np.zeros(3)
Obj0 = f(x0)

### Objective function
def Obj(x):
    y = f(x)/Obj0  #  (Rosenbrock function + x2)**2
    print(f'Obj: {y}, x: {x}')
    return y


### Inequality Constraint
def Cons1(x): # srqt(x1**2+x2**2-x1*x2+3*x3**2) < C --> 1-sqrt(...)/C >0
    C = 4.3
    con = 1-np.sqrt(x[0]**2 + x[1]**2 + 3*x[2]**2 - x[0]*x[1])/C
    print(f'con: {con}, x: {x}')
    return con

cons = {'type':'ineq',
        'fun':lambda x: Cons1(x)}

### Callback function
XX = [x0.copy()]
FF = [Obj(x0)]
def CB(x):
    print("In callback")
    global XX, FF
    XX += [x.copy()]
    FF += [Obj(x)]
    print(f'CB: {FF[-1]}, x: {x}')

    
### Bounds
boundmin = -5 * np.ones_like(x0)
boundmax = 5 * np.ones_like(x0)
bounds = spo.Bounds(boundmin, boundmax)

#%% OPTIMIZATION

# res = spo.minimize(Obj, x0, method='SLSQP',constraints=[cons], options={'ftol': 1e-6,'maxiter': 100,'eps':1.e-6},bounds=bounds)
res = spo.minimize(Obj, x0, callback=CB, method='SLSQP',constraints=[cons], options={'ftol': 1e-6, 'disp': True,'maxiter': 100, 'iprint': 2,'eps':1.e-6},bounds=bounds)
print(res)
```
I also modified the slsqp.py code (from master), adding a little debug information that might shed light on this. I added the line `print(f'BEFORE CALLBACK, x: {x}, mode: {mode}, majiter increased: {majiter > majiter_prev}')` just after the slsqp call, and `print(f'START OF LOOP, x: {x}, mode: {mode}')`

```
Obj: 1.0, x: [0. 0. 0.]
con: 1.0, x: [0. 0. 0.]
Obj: 1.0, x: [0. 0. 0.]
Obj: 0.999992589510571, x: [1.e-06 0.e+00 0.e+00]
Obj: 0.9999840001061049, x: [0.e+00 1.e-06 0.e+00]
Obj: 1.000000421052676, x: [0.e+00 0.e+00 1.e-06]
  NIT    FC           OBJFUN            GNORM
START OF LOOP, x: [0. 0. 0.], mode: 0
con: 1.0, x: [0. 0. 0.]
con: 1.0, x: [0. 0. 0.]
con: 0.9999997674418605, x: [1.e-06 0.e+00 0.e+00]
con: 0.9999997674418605, x: [0.e+00 1.e-06 0.e+00]
con: 0.9999995971974867, x: [0.e+00 0.e+00 1.e-06]
BEFORE CALLBACK, x: [ 5.          5.         -3.29089653], mode: 1, majiter increased: True
In callback
Obj: 48554370.97058353, x: [ 5.          5.         -3.29089653]
CB: 48554370.97058353, x: [ 5.          5.         -3.29089653]
    1     4     1.000000E+00     1.763772E+01
START OF LOOP, x: [ 5.          5.         -3.29089653], mode: 1
Obj: 48554370.97058353, x: [ 5.          5.         -3.29089653]
con: -0.763306054653021, x: [ 5.          5.         -3.29089653]
BEFORE CALLBACK, x: [ 0.5         0.5        -0.32908965], mode: 1, majiter increased: False
START OF LOOP, x: [ 0.5         0.5        -0.32908965], mode: 1
Obj: 1.9428899198715892, x: [ 0.5         0.5        -0.32908965]
con: 0.8236693945346979, x: [ 0.5         0.5        -0.32908965]
BEFORE CALLBACK, x: [ 0.23156495  0.23156495 -0.15241126], mode: 1, majiter increased: False
START OF LOOP, x: [ 0.23156495  0.23156495 -0.15241126], mode: 1
Obj: 0.0570541151358387, x: [ 0.23156495  0.23156495 -0.15241126]
con: 0.9183360250926375, x: [ 0.23156495  0.23156495 -0.15241126]
BEFORE CALLBACK, x: [ 0.23156495  0.23156495 -0.15241126], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.23156495  0.23156495 -0.15241126], mode: -1
Obj: 0.057053818030319316, x: [ 0.23156595  0.23156495 -0.15241126]
Obj: 0.05705573575688781, x: [ 0.23156495  0.23156595 -0.15241126]
Obj: 0.057054014563237755, x: [ 0.23156495  0.23156495 -0.15241226]
con: 0.9183360250926375, x: [ 0.23156495  0.23156495 -0.15241126]
con: 0.9183359484135263, x: [ 0.23156595  0.23156495 -0.15241126]
con: 0.9183359484135263, x: [ 0.23156495  0.23156595 -0.15241126]
con: 0.9183357222817453, x: [ 0.23156495  0.23156495 -0.15241226]
BEFORE CALLBACK, x: [ 1.12042018 -0.2898318  -0.70992618], mode: 1, majiter increased: True
In callback
Obj: 1201.3100512024678, x: [ 1.12042018 -0.2898318  -0.70992618]
CB: 1201.3100512024678, x: [ 1.12042018 -0.2898318  -0.70992618]
    2    10     5.705412E-02     1.650697E+00
START OF LOOP, x: [ 1.12042018 -0.2898318  -0.70992618], mode: 1
Obj: 1201.3100512024678, x: [ 1.12042018 -0.2898318  -0.70992618]
con: 0.5855462402319677, x: [ 1.12042018 -0.2898318  -0.70992618]
BEFORE CALLBACK, x: [ 0.32045047  0.17942527 -0.20816275], mode: 1, majiter increased: False
START OF LOOP, x: [ 0.32045047  0.17942527 -0.20816275], mode: 1
Obj: 0.003668436373208541, x: [ 0.32045047  0.17942527 -0.20816275]
con: 0.8940953085363905, x: [ 0.32045047  0.17942527 -0.20816275]
BEFORE CALLBACK, x: [ 0.32045047  0.17942527 -0.20816275], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.32045047  0.17942527 -0.20816275], mode: -1
Obj: 0.003668373184316397, x: [ 0.32045147  0.17942527 -0.20816275]
Obj: 0.003668512459063848, x: [ 0.32045047  0.17942627 -0.20816275]
Obj: 0.003668410871097835, x: [ 0.32045047  0.17942527 -0.20816375]
con: 0.8940953085363905, x: [ 0.32045047  0.17942527 -0.20816275]
con: 0.894095190703264, x: [ 0.32045147  0.17942527 -0.20816275]
con: 0.894095298731082, x: [ 0.32045047  0.17942627 -0.20816275]
con: 0.8940949896231515, x: [ 0.32045047  0.17942527 -0.20816375]
BEFORE CALLBACK, x: [ 0.33963024  0.17506864 -0.20536555], mode: 1, majiter increased: True
In callback
Obj: 0.002510016861156605, x: [ 0.33963024  0.17506864 -0.20536555]
CB: 0.002510016861156605, x: [ 0.33963024  0.17506864 -0.20536555]
    3    15     3.668436E-03     1.021384E-01
START OF LOOP, x: [ 0.33963024  0.17506864 -0.20536555], mode: 1
Obj: 0.002510016861156605, x: [ 0.33963024  0.17506864 -0.20536555]
con: 0.8926537829151637, x: [ 0.33963024  0.17506864 -0.20536555]
BEFORE CALLBACK, x: [ 0.33963024  0.17506864 -0.20536555], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.33963024  0.17506864 -0.20536555], mode: -1
Obj: 0.002509973998003043, x: [ 0.33963124  0.17506864 -0.20536555]
Obj: 0.002510040371486993, x: [ 0.33963024  0.17506964 -0.20536555]
Obj: 0.0025099957664352526, x: [ 0.33963024  0.17506864 -0.20536655]
con: 0.8926537829151637, x: [ 0.33963024  0.17506864 -0.20536555]
con: 0.8926536559037448, x: [ 0.33963124  0.17506864 -0.20536555]
con: 0.8926537802680772, x: [ 0.33963024  0.17506964 -0.20536555]
con: 0.892653472512363, x: [ 0.33963024  0.17506864 -0.20536655]
BEFORE CALLBACK, x: [ 0.36754679  0.16918756 -0.202847  ], mode: 1, majiter increased: True
In callback
Obj: 0.0017380043929366344, x: [ 0.36754679  0.16918756 -0.202847  ]
CB: 0.0017380043929366344, x: [ 0.36754679  0.16918756 -0.202847  ]
    4    19     2.510017E-03     5.324446E-02
START OF LOOP, x: [ 0.36754679  0.16918756 -0.202847  ], mode: 1
Obj: 0.0017380043929366344, x: [ 0.36754679  0.16918756 -0.202847  ]
con: 0.8896949890503334, x: [ 0.36754679  0.16918756 -0.202847  ]
BEFORE CALLBACK, x: [ 0.36754679  0.16918756 -0.202847  ], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.36754679  0.16918756 -0.202847  ], mode: -1
Obj: 0.0017379856200677923, x: [ 0.36754779  0.16918756 -0.202847  ]
Obj: 0.0017379732046138544, x: [ 0.36754679  0.16918856 -0.202847  ]
Obj: 0.0017379868395578572, x: [ 0.36754679  0.16918756 -0.202848  ]
con: 0.8896949890503334, x: [ 0.36754679  0.16918756 -0.202847  ]
con: 0.8896948503164139, x: [ 0.36754779  0.16918756 -0.202847  ]
con: 0.8896949962016222, x: [ 0.36754679  0.16918856 -0.202847  ]
con: 0.8896946906782578, x: [ 0.36754679  0.16918756 -0.202848  ]
BEFORE CALLBACK, x: [ 0.3753971   0.16742154 -0.2058764 ], mode: 1, majiter increased: True
In callback
Obj: 0.0016320928103264738, x: [ 0.3753971   0.16742154 -0.2058764 ]
CB: 0.0016320928103264738, x: [ 0.3753971   0.16742154 -0.2058764 ]
    5    23     1.738004E-03     4.041353E-02
START OF LOOP, x: [ 0.3753971   0.16742154 -0.2058764 ], mode: 1
Obj: 0.0016320928103264738, x: [ 0.3753971   0.16742154 -0.2058764 ]
con: 0.8876817208318264, x: [ 0.3753971   0.16742154 -0.2058764 ]
BEFORE CALLBACK, x: [ 0.3753971   0.16742154 -0.2058764 ], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.3753971   0.16742154 -0.2058764 ], mode: -1
Obj: 0.0016320806805941114, x: [ 0.3753981   0.16742154 -0.2058764 ]
Obj: 0.001632047420331996, x: [ 0.3753971   0.16742254 -0.2058764 ]
Obj: 0.0016320758001946126, x: [ 0.3753971   0.16742154 -0.2058774 ]
con: 0.8876817208318264, x: [ 0.3753971   0.16742154 -0.2058764 ]
con: 0.8876815803794142, x: [ 0.3753981   0.16742154 -0.2058764 ]
con: 0.8876817305953352, x: [ 0.3753971   0.16742254 -0.2058764 ]
con: 0.8876814234318413, x: [ 0.3753971   0.16742154 -0.2058774 ]
BEFORE CALLBACK, x: [ 0.38109956  0.16576976 -0.21554601], mode: 1, majiter increased: True
In callback
Obj: 0.001501273158151744, x: [ 0.38109956  0.16576976 -0.21554601]
CB: 0.001501273158151744, x: [ 0.38109956  0.16576976 -0.21554601]
    6    27     1.632093E-03     4.996726E-02
START OF LOOP, x: [ 0.38109956  0.16576976 -0.21554601], mode: 1
Obj: 0.001501273158151744, x: [ 0.38109956  0.16576976 -0.21554601]
con: 0.8839719087945954, x: [ 0.38109956  0.16576976 -0.21554601]
BEFORE CALLBACK, x: [ 0.38109956  0.16576976 -0.21554601], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.38109956  0.16576976 -0.21554601], mode: -1
Obj: 0.0015012665618374143, x: [ 0.38110056  0.16576976 -0.21554601]
Obj: 0.001501217602923664, x: [ 0.38109956  0.16577076 -0.21554601]
Obj: 0.001501256843978643, x: [ 0.38109956  0.16576976 -0.21554701]
con: 0.8839719087945954, x: [ 0.38109956  0.16576976 -0.21554601]
con: 0.8839717697899134, x: [ 0.38110056  0.16576976 -0.21554601]
con: 0.883971920344883, x: [ 0.38109956  0.16577076 -0.21554601]
con: 0.8839716073818369, x: [ 0.38109956  0.16576976 -0.21554701]
BEFORE CALLBACK, x: [ 0.38883288  0.16231312 -0.25243458], mode: 1, majiter increased: True
In callback
Obj: 0.0011354092149332155, x: [ 0.38883288  0.16231312 -0.25243458]
CB: 0.0011354092149332155, x: [ 0.38883288  0.16231312 -0.25243458]
    7    31     1.501273E-03     5.827561E-02
START OF LOOP, x: [ 0.38883288  0.16231312 -0.25243458], mode: 1
Obj: 0.0011354092149332155, x: [ 0.38883288  0.16231312 -0.25243458]
con: 0.8714406634707006, x: [ 0.38883288  0.16231312 -0.25243458]
BEFORE CALLBACK, x: [ 0.38883288  0.16231312 -0.25243458], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.38883288  0.16231312 -0.25243458], mode: -1
Obj: 0.0011354112152280557, x: [ 0.38883388  0.16231312 -0.25243458]
Obj: 0.001135342975587375, x: [ 0.38883288  0.16231412 -0.25243458]
Obj: 0.0011353950272580452, x: [ 0.38883288  0.16231312 -0.25243558]
con: 0.8714406634707006, x: [ 0.38883288  0.16231312 -0.25243458]
con: 0.8714405340350131, x: [ 0.38883388  0.16231312 -0.25243458]
con: 0.8714406769759546, x: [ 0.38883288  0.16231412 -0.25243458]
con: 0.8714403448823539, x: [ 0.38883288  0.16231312 -0.25243558]
BEFORE CALLBACK, x: [ 0.39326544  0.15696024 -0.33878889], mode: 1, majiter increased: True
In callback
Obj: 0.00048302609173833097, x: [ 0.39326544  0.15696024 -0.33878889]
CB: 0.00048302609173833097, x: [ 0.39326544  0.15696024 -0.33878889]
    8    35     1.135409E-03     6.777125E-02
START OF LOOP, x: [ 0.39326544  0.15696024 -0.33878889], mode: 1
Obj: 0.00048302609173833097, x: [ 0.39326544  0.15696024 -0.33878889]
con: 0.8419458295798001, x: [ 0.39326544  0.15696024 -0.33878889]
BEFORE CALLBACK, x: [ 0.39326544  0.15696024 -0.33878889], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.39326544  0.15696024 -0.33878889], mode: -1
Obj: 0.00048303290246496625, x: [ 0.39326644  0.15696024 -0.33878889]
Obj: 0.0004829698471156976, x: [ 0.39326544  0.15696124 -0.33878889]
Obj: 0.0004830168379491466, x: [ 0.39326544  0.15696024 -0.33878989]
con: 0.8419458295798001, x: [ 0.39326544  0.15696024 -0.33878889]
con: 0.8419457218658046, x: [ 0.39326644  0.15696024 -0.33878889]
con: 0.8419458431548366, x: [ 0.39326544  0.15696124 -0.33878889]
con: 0.8419454817973121, x: [ 0.39326544  0.15696024 -0.33878989]
BEFORE CALLBACK, x: [ 0.39083635  0.15190847 -0.4462136 ], mode: 1, majiter increased: True
In callback
Obj: 3.3827150877151484e-05, x: [ 0.39083635  0.15190847 -0.4462136 ]
CB: 3.3827150877151484e-05, x: [ 0.39083635  0.15190847 -0.4462136 ]
    9    39     4.830261E-04     5.740624E-02
START OF LOOP, x: [ 0.39083635  0.15190847 -0.4462136 ], mode: 1
Obj: 3.3827150877151484e-05, x: [ 0.39083635  0.15190847 -0.4462136 ]
con: 0.803522419133084, x: [ 0.39083635  0.15190847 -0.4462136 ]
BEFORE CALLBACK, x: [ 0.39083635  0.15190847 -0.4462136 ], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.39083635  0.15190847 -0.4462136 ], mode: -1
Obj: 3.382963900237953e-05, x: [ 0.39083735  0.15190847 -0.4462136 ]
Obj: 3.3810250980550114e-05, x: [ 0.39083635  0.15190947 -0.4462136 ]
Obj: 3.382470203249624e-05, x: [ 0.39083635  0.15190847 -0.4462146 ]
con: 0.803522419133084, x: [ 0.39083635  0.15190847 -0.4462136 ]
con: 0.8035223324571219, x: [ 0.39083735  0.15190847 -0.4462136 ]
con: 0.8035224311096207, x: [ 0.39083635  0.15190947 -0.4462136 ]
con: 0.8035220506528303, x: [ 0.39083635  0.15190847 -0.4462146 ]
BEFORE CALLBACK, x: [ 0.3893112   0.15045276 -0.48066176], mode: 1, majiter increased: True
In callback
Obj: 1.3905571096670256e-07, x: [ 0.3893112   0.15045276 -0.48066176]
CB: 1.3905571096670256e-07, x: [ 0.3893112   0.15045276 -0.48066176]
   10    43     3.382715E-05     1.725671E-02
START OF LOOP, x: [ 0.3893112   0.15045276 -0.48066176], mode: 1
Obj: 1.3905571096670256e-07, x: [ 0.3893112   0.15045276 -0.48066176]
con: 0.7908613721631034, x: [ 0.3893112   0.15045276 -0.48066176]
BEFORE CALLBACK, x: [ 0.3893112   0.15045276 -0.48066176], mode: -1, majiter increased: False
START OF LOOP, x: [ 0.3893112   0.15045276 -0.48066176], mode: -1
Obj: 1.3922191448547805e-07, x: [ 0.3893122   0.15045276 -0.48066176]
Obj: 1.3794674728490318e-07, x: [ 0.3893112   0.15045376 -0.48066176]
Obj: 1.3889874402723325e-07, x: [ 0.3893112   0.15045276 -0.48066276]
con: 0.7908613721631034, x: [ 0.3893112   0.15045276 -0.48066176]
con: 0.7908612909405973, x: [ 0.3893122   0.15045276 -0.48066176]
con: 0.7908613835938374, x: [ 0.3893112   0.15045376 -0.48066176]
con: 0.790860999265386, x: [ 0.3893112   0.15045276 -0.48066276]
BEFORE CALLBACK, x: [ 0.3893112   0.15045276 -0.48066176], mode: 0, majiter increased: True
In callback
Obj: 1.3905571096670256e-07, x: [ 0.3893112   0.15045276 -0.48066176]
CB: 1.3905571096670256e-07, x: [ 0.3893112   0.15045276 -0.48066176]
   11    47     1.390557E-07     1.132282E-03
Optimization terminated successfully    (Exit mode 0)
            Current function value: 1.3905571096670256e-07
            Iterations: 11
            Function evaluations: 47
            Gradient evaluations: 11
     fun: 1.3905571096670256e-07
     jac: array([ 0.0001662 , -0.00110896,  0.00015697])
 message: 'Optimization terminated successfully'
    nfev: 47
     nit: 11
    njev: 11
  status: 0
 success: True
       x: array([ 0.3893112 ,  0.15045276, -0.48066176])
```

The relevant slsqp is [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L407) to [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L478).
On successive iterations the code can either call objective and constraint evaluation (`mode in {0, 1}` followed by gradient evaluation (`mode in {-1, 0}`). From the detailed trace above there are several objective calls before the print on L404. Those objective calls come from setting up the ScalarFunction, and an initial gradient calculation.
The first time through the loop the mode = 0, so there are evaluations of the objective and constraint, as well as the gradients. At the end of that first loop the call to slsqp changes the `mode` to 1 and completes a major iteration so there is a callback initiated. The call to slsqp has changed the suggested solution, which is why the callback has a high objective value.
The second loop has mode = 1, evaluating objective and constraint. The constraint is also violated, `majiter` does not increase so there is no callback.
The third and fourth loops have mode = 1, evaluating the gradients, but this time the constraints are satisfied. `majiter` does not increase so there are no callbacks
Then slsqp changes the mode to -1 to evaluate gradients for the fifth loop.
The sixth loop changes the mode to 1, and also increase `majiter` so the `callback` is called.

The callback is only triggered when `majiter` increases, from inspection of the trace this only occurs when `mode is 1` at the end of the loop, but not on every instance when mode is 1. 

What's important to note here is that the `slsqp` call on L459 modifies the `x` solution vector, in between any locations where your objective is called and the callback. This is why the callback doesn't have the same `x` as when your objective is called. Plus, your objective could be called with many different `x` to evaluate gradients when `mode is -1`.
 You should therefore regard the `x` in the callback as the evolving solution vector, it's not the intermediate values of `x` in your objective that you want to be recording.

If the `slsqp` call sets mode to 0 during any iteration the loop will terminate. In some circumstances this may be with `majiter` being incremented, and sometimes not. This means that the callback may or may not be called in the last iteration. This is why the last value of `x` in the final callback may be different to OptimizeResult.x.

The value of OBJFUN printed by slsqp is stale at the time it's printed. `fx` is updated at the start of the loop (if mode is 1), but there is a call to `slsqp` before the callback and the printing of OBJFUN. This changes the `x` array between the two. Therefore, OBJFUN and `func(x)` don't tally at the time of printing, but the value of OBJFUN is correct at the start of the loop.

**Comment by Mocca25** [comment]
Thanks a lot for your detailed answer ! I think it's clearer in my mind. If I understand it correctly, the value displayed by slsqp.py at iteration n is the last function value actually computed by the algorithm (and that is not used for numerical gradient evaluation), but then the ```slsqp``` function can return a different ```x``` which generally seems to be the first ```x``` of the iteration n+1, and the callback function called at iteration n uses this new ```x```.

I'm not sure what you mean by "evolving solution vector". Do you mean I shouldn't be using a callback function, since it is evaluated with an untested ```x``` that is more likely to violate the constraints, and that is not representative of the algorithm convergence ?

Thanks again.

**Comment by andyfaff** [comment]
> the value displayed by slsqp.py at iteration n is the last function value actually computed by the algorithm (and that is not used for numerical gradient evaluation)

I think so. The value of `fx` (OBJFUN) printed in *major* iteration `n` does not correspond to `fun(x)` at the time the callback is triggered. This is because there is an intervening `slsqp` call, The following output was taken from your example where I print out f(x) in the callback. There is a 1 iteration lag between OBJFUN/`fx` and `fun(x)` in the callback. Even then the two don't have that lag at low iteration numbers because the first few iterations don't comply with the constraints.
```
 NIT    FC           OBJFUN            GNORM
Callback f(x): 48554370.97058353
    1     4     1.000000E+00     1.763772E+01
Callback f(x): 1201.3100512024678
    2    10     5.705412E-02     1.650697E+00
Callback f(x): 0.002510016861156605
    3    15     3.668436E-03     1.021384E-01
Callback f(x): 0.0017380043929366344
    4    19     2.510017E-03     5.324446E-02
Callback f(x): 0.0016320928103264738
    5    23     1.738004E-03     4.041353E-02
Callback f(x): 0.001501273158151744
    6    27     1.632093E-03     4.996726E-02
Callback f(x): 0.0011354092149332155
    7    31     1.501273E-03     5.827561E-02
Callback f(x): 0.00048302609173833097
    8    35     1.135409E-03     6.777125E-02
Callback f(x): 3.3827150877151484e-05
    9    39     4.830261E-04     5.740624E-02
Callback f(x): 1.3905571096670256e-07
   10    43     3.382715E-05     1.725671E-02
Callback f(x): 1.3905571096670256e-07
   11    47     1.390557E-07     1.132282E-03
```
I fix this lag in #11495.
I think you can still use the callback, it's representative of convergence. It's just that the `x` array is correct at time of the callback, but the OBJFUN print lags behind by one call to `slsqp`.
